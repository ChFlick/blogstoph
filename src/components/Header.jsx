import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';

const DEFAULT_HEADER_CLASSES = 'header header--fixed-top';

export class Header extends React.Component {
    state = {
        headerClasses: DEFAULT_HEADER_CLASSES
    }

    componentDidMount = () => {
        window.addEventListener('scroll', this.handleScroll);
    };

    componentWillUnmount = () => {
        window.removeEventListener('scroll', this.handleScroll);
    };

    handleScroll = () => {
        const scrollPosPrev = this.state.scrollPos;
        const scrollPos = window.scrollY;
        let headerClasses = 'header header--fixed-top';

        if (scrollPos > 80) {
            headerClasses = 'header header--fixed';
            if (scrollPos < scrollPosPrev) {
                headerClasses += ' header--visible';
            }

            this.setState(() => ({ headerClasses }));
        } else if (scrollPos === 0) {
            this.setState(() => ({ headerClasses: DEFAULT_HEADER_CLASSES }));
        }

        this.setState(() => ({ scrollPos }));
    };

    createAreaSwitchButton = () =>
        this.props.private ?
            <Link className="button button--link button--header" to="/dashboard">Visitor Area</Link> :
            <Link className="button button--link button--header" to="/editor/dashboard">Editor Area</Link>;


    render() {
        return (
            <header id="header" className={this.state.headerClasses} onScroll={this.handleScroll}>
                <div className="content-container">
                    <div className="header__content">
                        <Link className="header__title" to="/dashboard">
                            <h1>Blogstoph</h1>
                        </Link>
                        {this.props.isAuthenticated ? (
                            <div>
                                {this.createAreaSwitchButton()}
                                <button className="button button--link button--header" onClick={this.props.startLogout}>Logout</button>
                            </div>
                        ) : (
                                <Link className="button button--link" to="/login">Login</Link>
                            )}
                    </div>
                </div>
            </header>
        );
    };
};

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
});

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
