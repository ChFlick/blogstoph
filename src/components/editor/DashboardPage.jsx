import React from 'react';

const EditorDashboardPage = () => (
    <div>
        Editor Dasboard page content
    </div>
);

export default EditorDashboardPage;
