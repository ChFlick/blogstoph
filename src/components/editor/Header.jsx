import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../../actions/auth';


export const EditorHeader = ({ startLogout }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/editor/dashboard">
                    <h1>Blogstoph</h1>
                </Link>
                <button className="button button--link" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(EditorHeader);
