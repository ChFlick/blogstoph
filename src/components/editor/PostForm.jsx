import React from 'react';
import moment from 'moment';
import Markdown from 'react-remarkable';
import { SingleDatePicker } from 'react-dates';

class PostForm extends React.Component {
    constructor(props) {
        super(props);

        const post = props.post || {};
        this.state = {
            title: post.title || '',
            content: post.content || '',
            author: post.author || '',
            date: moment(post.date) || moment(),
            dateFocused: false,
            published: post.published || false,
            error: ''
        };
    }

    onSubmit = (e) => {
        e.preventDefault();

        const { title, content, author, date, published } = this.state;

        if (published && (!title || !content || !author || !date)) {
            this.setState(() => ({
                error: 'Please provide all information when publishing a post.'
            }));
        } else if (!title) {
            this.setState(() => ({
                error: 'Please provide a title to identify the post.'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));

            this.props.onSubmit({
                title,
                content,
                author,
                date: date.valueOf(),
                published
            });
        }
    };

    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({
            title
        }));
    };

    onContentChange = (e) => {
        const content = e.target.value;
        this.setState(() => ({
            content
        }));
    };

    onAuthorChange = (e) => {
        const author = e.target.value;
        this.setState(() => ({
            author
        }));
    };

    onDateChange = (date) => {
        this.setState(() => ({ date }));
    };

    onPublishedChange = (e) => {
        const published = e.target.checked;
        this.setState(() => ({
            published
        }));
    };

    createTitleInput = () => (
        <input
            type="text"
            placeholder="Title *"
            onChange={this.onTitleChange}
            value={this.state.title}
            className="text-input"
        />
    );

    createContentInput = () => (
        <textarea
            type="text"
            placeholder="Content"
            onChange={this.onContentChange}
            value={this.state.content}
            className="textarea"
        />
    );

    createAuthorInput = () => (
        <input
            type="text"
            placeholder="Author"
            onChange={this.onAuthorChange}
            value={this.state.author}
            className="text-input"
        />
    );

    createDatePicker = () => (
        <SingleDatePicker
            date={this.state.date}
            onDateChange={this.onDateChange}
            focused={this.state.dateFocused}
            onFocusChange={({ focused }) => this.setState({ dateFocused: focused })}
            numberOfMonths={1}
            displayFormat="DD.MM.YYYY"
        />
    );

    createPublishedCheckbox = () => (
        <div>
            <div className="pretty p-svg p-curve p-smooth">
                <input
                    type="checkbox"
                    onChange={this.onPublishedChange}
                    checked={this.state.published}
                />
                <div className="state p-success">
                    <svg className="svg svg-icon" viewBox="0 0 20 20">
                        <path d="M7.629,14.566c0.125,0.125,0.291,0.188,0.456,0.188c0.164,0,
                        0.329-0.062,0.456-0.188l8.219-8.221c0.252-0.252,
                        0.252-0.659,0-0.911c-0.252-0.252-0.659-0.252-0.911,
                        0l-7.764,7.763L4.152,9.267c-0.252-0.251-0.66-0.251-0.911,
                        0c-0.252,0.252-0.252,0.66,0,0.911L7.629,14.566z" style={{stroke: 'white', fill: 'white'}}></path>
                    </svg>
                    <label>Published</label>
                </div>
            </div>
        </div>
    );

    createButtons = () => (
        <div>
            <button
                className="button button--ingroup"
                type="button"
                onClick={this.props.onBack}
            >
                Back
            </button>
            <button className="button button--ingroup">
                Save Post
            </button>
        </div>
    );

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                {this.createTitleInput()}
                {this.createContentInput()}
                <Markdown source={this.state.content} />
                {this.createAuthorInput()}
                {this.createDatePicker()}
                {this.createPublishedCheckbox()}
                {this.createButtons()}
            </form>
        );
    };
}

export default PostForm;
