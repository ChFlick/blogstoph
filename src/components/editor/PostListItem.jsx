import React from 'react';
import { Link } from 'react-router-dom';

const PostListItem = (props) => (
    <div>
        <Link to={`post/${props.post.id}`}>{props.post.title}<br /></Link>
    </div>
);

export default PostListItem;
