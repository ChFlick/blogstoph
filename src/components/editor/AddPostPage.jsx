import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import { startAddPost } from '../../actions/posts';
import PostForm from './PostForm';

export class AddPostPage extends React.Component {
    goBack = () => {
        this.props.history.push('/editor/dashboard');
    }

    onSubmit = (post) => {
        this.props.addPost(post);
        this.goBack();
    };

    render() {
        return (
            <Fragment>
                <h1>Add Post</h1>
                <PostForm onSubmit={this.onSubmit} onBack={this.goBack} />
            </Fragment>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    addPost: (post) => dispatch(startAddPost(post))
});

export default connect(undefined, mapDispatchToProps)(AddPostPage);
