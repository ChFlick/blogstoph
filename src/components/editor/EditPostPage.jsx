import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';

import { startEditPost, startRemovePost } from '../../actions/posts';
import PostForm from './PostForm';
import { getPostByIdOrTitle } from '../../selectors/posts';

export class EditPostPage extends React.Component {
    state = {
        modalIsOpen: false
    };

    openModal = () => {
        this.setState(() => ({ modalIsOpen: true }));
    }

    closeModal = () => {
        this.setState(() => ({ modalIsOpen: false }));
    }

    goBack = () => {
        this.props.history.push('/editor/dashboard');
    };

    onSubmit = (post) => {
        this.props.editPost(this.props.post.id, post);
        this.goBack();
    };

    onRemove = () => {
        this.props.removePost(this.props.post);
        this.goBack();
    };

    pageHeader = (
        <header className="page-header">
            <div className="content-container">
                <h1 className="page-header__title ">Edit Post</h1>
            </div>
        </header>
    );

    createPageContent = () => (
        <section className="content-container">
            <PostForm
                post={this.props.post}
                onSubmit={this.onSubmit}
                onBack={this.goBack} />
            <button
                className="button"
                type="button"
                onClick={this.openModal}
            >
                Remove Post
            </button>
        </section>
    );

    createModal = () => (
        <Modal
            isOpen={this.state.modalIsOpen}
            onRequestClose={this.closeModal}
            contentLabel="Remove Post"
        >
            <section>
                <h2>Remove Post</h2>
                <p>Are you sure you want to remove the post?</p>
                <button onClick={this.closeModal}>Cancel</button>
                <button onClick={this.onRemove}>Ok</button>
            </section>
        </Modal>
    );

    render() {
        return (
            <Fragment>
                {this.pageHeader}
                {this.createPageContent()}
                {this.createModal()}
            </Fragment>
        );
    };
};

const mapStateToProps = (state, props) => ({
    post: getPostByIdOrTitle(state.posts, props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    editPost: (id, post) => dispatch(startEditPost(id, post)),
    removePost: (id) => dispatch(startRemovePost(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);
