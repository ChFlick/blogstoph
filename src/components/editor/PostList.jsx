import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import PostListItem from './PostListItem';

export const PostList = (props) => (
    <Fragment>
        {
            props.posts.length === 0 ? (
                <p>There are no posts available</p>
            ) : (
                props.posts.map((post) => (
                    <PostListItem post={post} key={post.id} />
                ))
            )
        }
    </Fragment>
);

const mapStateToProps = (state) => ({
    posts: state.posts || []
});

export default connect(mapStateToProps)(PostList);
