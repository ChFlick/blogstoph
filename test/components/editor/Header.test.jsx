import React from 'react';
import { shallow } from 'enzyme';

import { EditorHeader } from '../../../src/components/editor/Header';


test('should render EditorHeader correctly', () => {
    const wrapper = shallow(<EditorHeader startLogout={() => {}} />);

    expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<EditorHeader startLogout={startLogout} />);

    wrapper.find('button').simulate('click');

    expect(startLogout).toHaveBeenCalled();
});
