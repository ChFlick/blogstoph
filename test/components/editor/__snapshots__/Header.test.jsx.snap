// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditorHeader correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/editor/dashboard"
      >
        <h1>
          Blogstoph
        </h1>
      </Link>
      <button
        className="button button--link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
